storage: report failed raft command application

tbg opened the pull request:
Previously, if `applyRaftCommand` returned an error, it would mark the replica
as corrupt but then go on and execute the side effects and potentially the
assertions in `assertState()`. These were then likely to fail and return a
misleading error, as likely seen in #16004.

Instead, cause a fatal error right when observing the error, and potentially
capture the root cause on sentry.io. The (perhaps too optimisic) expectation
is that after accounting these disk corruption/space errors, there will be
much fewer (possibly no) reports triggered by `assertState()`.

Touches #16004.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/19447)
<!-- Reviewable:end -->

nvanbenschoten commented:
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica.go, line 4464 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19447#-Kx8uiLW1uEbs1MXSBIs:-Kx8uiLW1uEbs1MXSBIt:b95gte1) ([raw file](https://github.com/cockroachdb/cockroach/blob/2613077f41fca134abe1945358b24126f125655f/pkg/storage/replica.go#L4464)):*
> ```Go
> 			// descriptive message so that we can differentiate the root causes.
> 			if err != nil {
> 				log.Errorf(ctx, "unable to update the state machine: %s", err)
> ```

Why add the `log.Error` before the `log.Fatal`? Is there a difference between them that I'm not aware of?

---

*[pkg/storage/replica.go, line 4481 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19447#-Kx8uxxtq3LC52EW-1fT:-Kx8uxxtq3LC52EW-1fU:bvxhue4) ([raw file](https://github.com/cockroachdb/cockroach/blob/2613077f41fca134abe1945358b24126f125655f/pkg/storage/replica.go#L4481)):*
> ```Go
> 		// interesting errors originate in applyRaftCommand, and they are
> 		// already handled above.
> 		pErr = r.maybeSetCorrupt(ctx, pErr)
> ```

What's the deal with this? Can we delete it? Can it be moved into the `TestingPostApplyFilter` block?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19447)*
<!-- Sent from Reviewable.io -->

tbg committed:
storage: report failed raft command application

Previously, if `applyRaftCommand` returned an error, it would mark the replica
as corrupt but then go on and execute the side effects and potentially the
assertions in `assertState()`. These were then likely to fail and return a
misleading error, as likely seen in #16004.

Instead, cause a fatal error right when observing the error, and potentially
capture the root cause on sentry.io. The (perhaps too optimisic) expectation
is that after accounting these disk corruption/space errors, there will be
much fewer (possibly no) reports triggered by `assertState()`.

Touches #16004.

tschottdorf commented:
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica.go, line 4464 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19447#-Kx8uiLW1uEbs1MXSBIs:-Kx8wGzVzFZuKG5dZWPe:b-ur7cdn) ([raw file](https://github.com/cockroachdb/cockroach/blob/2613077f41fca134abe1945358b24126f125655f/pkg/storage/replica.go#L4464)):*
Oops, yeah, that's subtle. Added comment:

> // Report the fatal error separately and only with the error, as that
// triggers an optimization for which we directly report the error to
// sentry (which in turn allows sentry to distinguish different error
// types).

---

*[pkg/storage/replica.go, line 4481 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19447#-Kx8uxxtq3LC52EW-1fT:-Kx8w0U8sIGG7S7SDdz7:b-ltrar5) ([raw file](https://github.com/cockroachdb/cockroach/blob/2613077f41fca134abe1945358b24126f125655f/pkg/storage/replica.go#L4481)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

What's the deal with this? Can we delete it? Can it be moved into the `TestingPostApplyFilter` block?
</blockquote></details>

The replica corruption handling has been around for a while, it's pretty half-assed but has enough pull that I don't want to rip it out during cherry-picking season (especially since we'll want something like it at some point, but not quite yet, so it's not clear whether to keep or not). I'd rather leave it for now and definitely in this PR.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19447)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19447#-:-Kx9ULRBymhrx92_nxiJ:bnfp4nl)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19447#-:-Kx9Xd98ykPM37KrRhhB:bnfp4nl)*
<!-- Sent from Reviewable.io -->
